tracker/progress.go

```diff
@@ -35,8 +35,22 @@ type Progress struct {
 	// entries with indices in (Match, Next) interval are already in flight.
 	//
 	// Invariant: 0 <= Match < Next.
+	// NB: it follows that Next >= 1.
 	Next uint64
 
+	// pendingCommit is the highest commit index <= Next-1 in flight to the
+	// follower.
+	//
+	// The actual in-flight commit index can be higher, but we track only up to
+	// Next-1, because higher indices can be ignored by the peer or reduced to
+	// Next-1 when all the in-flight entries are appended. When Next regresses,
```

Comment:
This is an incomplete leader-side copy of the follower logic that caps the commit index. We may pessimistically lower pendingCommit based on an update to Next, which would result in an additional message to convey an updated commit index. This PR tests only the reverse case, when the tracking of pendingCommit *prevents* the sending of a message. 

In general this feels fragile. The goal of my other comment was to get away from heuristics and logic that is duplicated between leader and follower side. I wanted some way for the follower to say "I believe the commit index is X" instead of the leader saying that "when Next updates, set Progress.pendingCommit".

Reply:
I agree, this logic would be more complete if the leader tracked the follower's commit index based on what the follower says.

This is similar to the `Match/Next` tracking for the log state:

1. `Match` tracks the log index that is stored durably on the follower (confirmed by a follower->leader message).
2. `Next` tracks the optimistic log index - all entries in `(Match, Next)` are in-flight to the follower.

For commit index, we should similarly have:

1. `Committed` index that is stored durably on the follower (confirmed by a follower->leader message).
2. `pendingCommit >= Commit` index that is in-flight to the follower.

This PR implements part (2), and improves status quo. Implementing part (1) requires a protocol change. This is captured by #131.

For the purposes of flow control improvements that I'm making, adding only part (2) for now is good enough.

Reply:
> We may pessimistically lower pendingCommit based on an update to Next, which would result in an additional message to convey an updated commit index.

Note that we send the actual commit index, not the capped `pendingCommit`. We cap `pendingCommit` only here in the tracker.

Reply:
I agree that this seems a bit fragile. It would be better for the follower to communicate its commit index. It isn't clear to me that there is enough of a marginal benefit here to justify an intermediate solution prior to a protocol change.

Is this change a hard requirement for the flow control work, or simply an opportunistic improvement?

Reply:
> Is this change a hard requirement for the flow control work, or simply an opportunistic improvement?

It is a requirement. We need to get rid of `sendIfEmpty` flag in `maybeSendAppend`, so that all the information necessary to send a message is consolidated in `Progress` struct. The decision-making for sending an empty `MsgApp` is currently not entirely encoded in `Progress`, and commit index bump is one of the cases we should "migrate" to `Progress`.

Reply:
I can move forward with the follower->leader commit index protocol change, since it seems like a common sentiment here. This should be a small backwards-compatible PR.

Reply:
I added the follower commit index tracking in #132 and rebased this PR on top. However, I don't see much value in doing both things now. The `pendingCommit` index is still necessary, and remained in its original form.

Reply:
I think we should revert this PR to the original form that was approved last week.

Reply:
There is one case in which tracking both `Commit` and `pendingCommit` can help.

When `Commit == log.lastIndex()`, we don't need to keep sending the empty `MsgApp` periodically. However, we can only take advantage of this case in PR #134.

Reply:
@erikgrinaker Are you happy to leave this PR as is? My assessment is that we need both parts (1) and (2), but (2) [this PR] is needed immediately. Also, (2) won't change after we add (1) - we still need to cap `pendingCommit` at `Next-1` etc.

I can maybe add some tests to check the invariants, but I already audited all the code that modifies `Next`, and made sure `pendingCommit < Next` invariant holds.

Reply:
>The pendingCommit index is still necessary, and remained in its original form. Only pendingCommit is needed to decide on the next empty MsgApp message, the Commit index of the follower is not used.

Why? I'm not seeing `pendingCommit` in #132. It seems like tracking the actual acknowledged commit index would remove some assumptions, and also allow eagerly retransmitting the commit index if the previous message may have been lost (which isn't the case with `pendingCommit`).

As I understand it, the main objective here is to make `maybeSendAppend` only depend on the progress state. Isn't a follower `Commit` field sufficient for this?

Reply:
Example why only `Commit` is not sufficient. Consider the case:
- leader log is at index 100, and commit index is 80
- follower's `Commit` is 50, and `Match` is 50

We send a `MsgApp` with entries [50,100] and commit 80. Now we need to pause sending any more messages. We will be sending periodic empty "probe" messages instead (for 2 reasons: `Match < raftLog.lastIndex` and `Commit < raftLog.committed`).

Say, the leader's commit index was updated to 100. We need to send it immediately, instead of waiting for the next "probing" attempt, because this index has a chance of applying (all entries up to 100 are in-flight too). To distinguish this case, we need the `pendingCommit < raftLog.committed` check (80 < 100). We can't use `Commit < raftLog.committed` check because we will end up spinning.

Reply:
> it would be nice if pendingCommit contained the actual commit index we sent (not the truncated one)

I'll consider it in #132. Not sure we should do this, because sending a commit index >= `Next` is unlikely to be applied. The leader's assumption is that the follower has all entries up to `Match`, and potentially soon will have the in-flight entries `(Match, Next)`. So if the leader's commit index is >= `Next`, it doesn't make much sense to send it more than once.

Reply:
> sending a commit index >= Next is unlikely to be applied

Why do we even send it in the first place then?

Reply:
Yes, but why don't we truncate it to Next?

Reply:
Because there is a chance that the follower is more up-to-date than `Next`. In this case, it will apply a higher commit index, and reply with a `MsgAppResp` that resets the leader->follower flow to a more up-to-date state. This is to address the "less likely", but still possible scenario.

Reply:
Seems like we shouldn't hide that fact in the field in that case, but 🤷

Reply:
You mean we can set `sentCommit` to the actual commit index we sent? We can do that, yes. Then the condition for sending eagerly would transform into:

```golang
raftLog.committed > pr.sentCommit && pr.sentCommit < pr.Next-1
```

Reply:
Yep. That seems clearer to me, but no strong opinion.

Reply:
Updated the second commit. Now we don't need the invariant. Plus, this further reduced unnecessary sends (see `testdata` files diffs).

Reply:
@erikgrinaker considering this resolved for now. Please comment on the updated PR if you have further suggestions.